thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/TriggerConditionGrouper.java

```diff
@@ -0,0 +1,283 @@
+   *
+   * Since the anomalies from the respective entities/metrics are merged
+   * before calling the grouper, we do not have to deal with overlapping
+   * anomalies within an entity/metric
+   *
+   * Core logic:
+   * Sort anomalies by start time and iterate through the list
+   * Push first anomaly to stack
+   * For each anomaly following, pop stack and compare with current anomaly
+   * 1. No overlap:
+   *    If the anomaly doesn't overlap, create and push new entity anomaly
+   * 2. Partial overlap:
+   *    If partial overlap, create new grouped entity anomaly and push to stack followed by current anomaly
+   * 3. Full overlap:
+   *    If full overlap, create new grouped entity anomaly and push to stack followed by current anomaly
+   * After the last anomaly, the stack will have one extra anomaly which needs to be popped out.
+   * Now the stack will hold all the grouped/entity anomalies
+   *
+   * Note the each time an entity copy is created, the children needs to be marked and set appropriately
+   *
+   */
+  private List<MergedAnomalyResultDTO> andGrouping(
+      List<MergedAnomalyResultDTO> anomalyListA, List<MergedAnomalyResultDTO> anomalyListB) {
+    Set<MergedAnomalyResultDTO> groupedAnomalies = new HashSet<>();
+    List<MergedAnomalyResultDTO> anomalies = mergeAndSortAnomalies(anomalyListA, anomalyListB);
+
+    Stack<MergedAnomalyResultDTO> anomalyStack = new Stack<>();
+    for (MergedAnomalyResultDTO anomaly : anomalies) {
+      if (anomalyStack.isEmpty()) {
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+        continue;
+      }
+
+      MergedAnomalyResultDTO anomalyStackTop = anomalyStack.pop();
+
+      if (anomaly.getStartTime() > anomalyStackTop.getEndTime()) {
+        // No overlap
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+      } else if (anomaly.getEndTime() > anomalyStackTop.getEndTime()) {
+        //Partial overlap
+        anomalyStackTop.setStartTime(anomaly.getStartTime());
+        setEntityChildMapping(anomalyStackTop, anomaly);
+
+        anomalyStack.push(anomalyStackTop);
+        anomalyStack.push(makeParentEntityAnomaly(anomaly));
+      } else {
+        // Complete overlap
+        MergedAnomalyResultDTO temp = makeEntityCopy(anomalyStackTop);
+
+        anomalyStackTop.setStartTime(anomaly.getStartTime());
+        anomalyStackTop.setEndTime(anomaly.getEndTime());
+        setEntityChildMapping(anomalyStackTop, anomaly);
+
+        anomalyStack.push(anomalyStackTop);
+        anomalyStack.push(temp);
+      }
+    }
+
+    // Pop the extra anomaly out
+    if (!anomalyStack.isEmpty()) {
+      anomalyStack.pop();
+    }
+
+    // Add all parent anomalies along with their children
+    groupedAnomalies.addAll(anomalyStack);
+    for (MergedAnomalyResultDTO anomaly : anomalyStack) {
+      groupedAnomalies.addAll(anomaly.getChildren());
+    }
+
+    return new ArrayList<>(groupedAnomalies);
```

Comment:
Could be  simplified as:
 List<MergedAnomalyResultDTO> anomalies = mergeAndSortAnomalies(anomalyListA, anomalyListB);
 if (anomalyListA.isEmpty() && anomalyListB.isEmpty()) return anomalies;

int i = 0;
 for (int i = 0; i < anomalies.size();  i++) {
       for (int j = i + 1; j < anomalies.size(); j++) {
           // if has overlap then output the overlap part
           if (anomalies.get(j). getStartTime() <= anomalies.get(i).getEndTime())) {
                MergedAnomalyResultDTO cur = makeParentEntityAnomaly(anomalies.get(i));
                cur. setEndTime(Math.min(cur. getEndTime()), anomalies.get(j). getEndTime()));
                cur. setStartTime(anomalies.get(j). getStartTime());
                setEntityChildMapping(cur, anomalies.get(i));
                anomalies.add(cur);
       }  else {
          break;
      }
  }
}
anomalies.add(cur);
return anomalies;

Reply:
updated accordingly